app/src/extra/java/org/wikipedia/push/WikipediaFirebaseMessagingService.kt

```diff
@@ -0,0 +1,172 @@
+        //remoteMessage.notification?.let {
+        //    ...
+        //}
+    }
+
+    // Called when a token is first generated for the app, or when a token is revoked and
+    // regenerated for security reasons.
+    override fun onNewToken(token: String) {
+        L.d("Received a new Firebase token...")
+
+        // As soon as we receive a new token, it's super important to save it in our Prefs, because
+        // another one might not be generated for a long time, and we should preserve it in case
+        // the subscription API happens to fail.
+        Prefs.setPushNotificationTokenOld(Prefs.getPushNotificationToken())
+        Prefs.setPushNotificationToken(token)
+        Prefs.setPushNotificationTokenSubscribed(false)
+
+        updateSubscription()
+    }
+
+    private fun handleCheckEcho() {
+        if (!Prefs.notificationPollEnabled()) {
+            return
+        }
+        NotificationPollBroadcastReceiver.pollNotifications(this)
+    }
+
+    companion object {
+        const val MESSAGE_TYPE_CHECK_ECHO = "checkEchoV1"
+        private const val SUBSCRIBE_RETRY_COUNT = 5
+        private const val UNSUBSCRIBE_RETRY_COUNT = 3
+
+        fun isUsingPush(): Boolean {
+            return GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(WikipediaApp.getInstance()) == ConnectionResult.SUCCESS
+                    && Prefs.getPushNotificationToken().isNotEmpty()
+                    && Prefs.isPushNotificationTokenSubscribed()
+        }
+
+        fun updateSubscription() {
+            if (!AccountUtil.isLoggedIn()) {
+                // Don't bother doing anything if the user is not logged in.
+                return
+            }
+
+            CsrfTokenClient(WikipediaApp.getInstance().wikiSite).request(false, object : CsrfTokenClient.Callback {
+                override fun success(token: String) {
+                    subscribeWithCsrf(token)
+                    setNotificationOptions(token)
+                }
+
+                override fun failure(t: Throwable) {
+                    L.e(t)
+                }
+
+                override fun twoFactorPrompt() {
+                    // ignore
+                }
+            })
+        }
+
+        private fun subscribeWithCsrf(csrfToken: String) {
+            if (Prefs.isPushNotificationTokenSubscribed() || Prefs.getPushNotificationToken().isEmpty()) {
+                // Don't do anything if the token is already subscribed, or if the token is empty.
+                return
+            }
+
+            val token = Prefs.getPushNotificationToken()
+            val oldToken = Prefs.getPushNotificationTokenOld()
+
+            // Make sure to unsubscribe the previous token, if any
+            if (oldToken.isNotEmpty()) {
+                if (oldToken != token) {
+                    ServiceFactory.get(WikipediaApp.getInstance().wikiSite).unsubscribePush(csrfToken, Prefs.getPushNotificationToken())
```

Comment:
A little bit confused here. Should we unsubscribe the `token` or `oldToken`?

Reply:
Whoops, fixed!